ee/test/test_migration_0012.py

```diff
@@ -0,0 +1,103 @@
+from django.db import transaction
+from django.db.models import Q
+
+from posthog.models import Tag as TagModel
+from posthog.models import TaggedItem as TaggedItemModel
+from posthog.models import Team
+from posthog.test.base import TestMigrations
+
+
+class TagsTestCase(TestMigrations):
+    migrate_from = "0011_add_tags_back"  # type: ignore
+    migrate_to = "0012_migrate_tags_v2"  # type: ignore
+
+    @property
+    def app(self):
+        return "ee"
+
+    def setUpBeforeMigration(self, apps):
+        EnterpriseEventDefinition = apps.get_model("ee", "EnterpriseEventDefinition")
+        EnterprisePropertyDefinition = apps.get_model("ee", "EnterprisePropertyDefinition")
+
+        # Setup
+        # apps.get_model("posthog", "Tag") doesn't work in setup because of a dependency issue
+        tag = TagModel.objects.create(name="existing tag", team_id=self.team.id)
+        self.event_definition = EnterpriseEventDefinition.objects.create(
+            team_id=self.team.id,
+            name="enterprise event",
+            deprecated_tags=["a", "b", "c", "a", "b", "existing tag", "", "  ", None],
+        )
+        self.property_definition_with_tags = EnterprisePropertyDefinition.objects.create(
+            team_id=self.team.id, name="property def with tags", deprecated_tags=["c", "d", "d", "existing tag"],
+        )
+        self.property_definition_without_tags = EnterprisePropertyDefinition.objects.create(
+            team_id=self.team.id, name="property def without tags",
+        )
+        TaggedItemModel.objects.create(tag=tag, property_definition_id=self.property_definition_with_tags.id)
+
+        # Setup for batched tags
+        self.team2 = Team.objects.create(
+            organization=self.organization,
+            api_token="token12345",
+            test_account_filters=[
+                {"key": "email", "value": "@posthog.com", "operator": "not_icontains", "type": "person"},
+            ],
+        )
+        self.team2_total_property_definitions = 2_300
+        tag2 = TagModel.objects.create(name="existing tag", team_id=self.team2.id)
+        with transaction.atomic():
+            for _tag in range(self.team2_total_property_definitions):
+                EnterprisePropertyDefinition.objects.create(
+                    name=f"batch_prop_{_tag}",
+                    team_id=self.team2.id,
+                    deprecated_tags=[_tag, "existing tag", "tag to create"],
+                )
+        TaggedItemModel.objects.create(
+            tag=tag2,
+            property_definition_id=EnterprisePropertyDefinition.objects.filter(team_id=self.team2.id).first().id,
+        )
+
+    def test_tags_migrated(self):
```

Comment:
Can we run the migration twice?

non-blocking: I'm not sure how that fits in with the `TestMigration` structure, but tbh I'd prefer to just perform the migrations explicitly in these tests rather implicitly somewhere in the class hierarchy

Reply:
I think this will require some custom boilerplate since `TestMigration` intercepts the one time `setUp` call to perform the migration. I agree it'd be nice to have separate setup, execute, and teardown methods for each test but I think it'll be fine for now.